### PR TITLE
Refactor upper envelope and write docstrings

### PR DESCRIPTION
**Pull request by MaxBlesch**
Refactor upper envelope step and write docstrings.

### CODE REVIEW

File: src/dcegm/fast_upper_envelope.py [hunk 1]
```diff
@@ -825,57 +668,555 @@ def _backward_scan(
     )
 
 
-def _evaluate_point_on_line(
-    x1: float, y1: float, x2: float, y2: float, point_to_evaluate: float
-) -> float:
-    """Evaluate a point on a line.
+def update_bools_and_idx_to_inspect(idx_to_inspect, update_idx, case_2, case_5):
```
Comment (segsell) [inline src/dcegm/fast_upper_envelope.py:671]: Can you check whether the docstring and arg definitions here are correct?

File: src/dcegm/fast_upper_envelope.py [hunk 2]
```diff
@@ -825,57 +668,555 @@ def _backward_scan(
...
     Args:
-        x1 (float): x coordinate of the first point.
-        y1 (float): y coordinate of the first point.
-        x2 (float): x coordinate of the second point.
-        y2 (float): y coordinate of the second point.
-        point_to_evaluate (float): The point to evaluate.
+        idx_to_inspect (int): Index of the point to be inspected in the current
+            iteration.
+        update_idx (bool): Indicator if the index should be updated.
+        case_2 (bool): Indicator if we have reached the final point on the grid.
```
Comment (segsell) [inline src/dcegm/fast_upper_envelope.py:682]: case_2 correct?

File: src/dcegm/fast_upper_envelope.py [hunk 3]
```diff
@@ -825,57 +668,555 @@ def _backward_scan(
...
+        case_5 (bool): Indicator if we are in the situation where we added the
```
Comment (segsell) [inline src/dcegm/fast_upper_envelope.py:683]: case_5 was a little unclear to me. This one should be updated.

File: src/dcegm/fast_upper_envelope.py [hunk 4]
```diff
@@ -825,57 +668,555 @@ def _backward_scan(
...
+            point that lies on the upper envelope (j).
+        last_point_was_intersect (bool): Indicator if the last point was an
+            intersection point.
+        is_final_point_on_grid (bool): Indicator if this is the final point on the grid.
+        jump_thresh (float): Jump detection threshold.
+
+    Returns:
+        tuple:
+
+        - cases (tuple): Tuple containing the indicators for the different cases.
```
Comment (segsell) [inline src/dcegm/fast_upper_envelope.py:890]: Add a concrete definition for each of the 6 cases here.

**Review by segsell — CHANGES_REQUESTED** [review]

**Review by segsell — APPROVED** [review]

File: src/dcegm/fast_upper_envelope.py (hunk 2 above)
Comment (MaxBlesch) [reply src/dcegm/fast_upper_envelope.py:682]: Yes

File: src/dcegm/fast_upper_envelope.py (hunk 3 above)
Comment (MaxBlesch) [reply src/dcegm/fast_upper_envelope.py:683]: It will. But at a later point!

File: src/dcegm/fast_upper_envelope.py (hunk 1 above)
Comment (MaxBlesch) [reply src/dcegm/fast_upper_envelope.py:671]: Yes, why not?

**Pull request merged**